### PR TITLE
Prepared for Serverless Spark cluster and job

### PR DESCRIPTION
**Pull request by wezhang**

### CODE REVIEW

File: Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/sdk/common/azure/serverless/AzureSparkServerlessCluster.java [hunk 1]
```diff
@@ -0,0 +1,496 @@
...
+import java.net.URI;
+import java.nio.charset.StandardCharsets;
+import java.util.Arrays;
+import java.util.List;
+import java.util.UUID;
+
+public class AzureSparkServerlessCluster extends SparkCluster
+                                         implements ProvisionableCluster,
+                                                    ServerlessCluster,
+        DestroyableCluster {
```
Comment (djlan) [inline Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/sdk/common/azure/serverless/AzureSparkServerlessCluster.java]: >DestroyableCluster [](start = 8, length = 18)

NIT, formatting

File: Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/spark/common/SparkBatchAzureSubmission.java [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+    public SparkBatchAzureSubmission setTenantId(String tid) {
+        this.tenantId = tid;
+
+        return this;
+    }
+
+    @Override
+    public void setCredentialsProvider(String username, String password) {
+        getCredentialsProvider().clear();
+    }
```
Comment (djlan) [inline Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/spark/common/SparkBatchAzureSubmission.java]: Username and password?

File: Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/spark/common/SparkBatchRemoteDebugJob.java [hunk 3]
```diff
@@ -206,4 +211,13 @@ static public SparkSubmissionParameter convertToDebugParameter(SparkSubmissionPa
...
     }
+
+    @Override
+    public boolean isRunning(@NotNull String state) {
+        try {
+            return getSparkDriverDebuggingPort().toBlocking().singleOrDefault(-1) > 0;
+        } catch (Exception e) {
+            return false;
+        }
+    }
```
Comment (djlan) [inline Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/spark/common/SparkBatchRemoteDebugJob.java]: This looks wield for isRunning call

File: Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/spark/common/SparkBatchJob.java [hunk 4]
```diff
@@ -636,38 +718,49 @@ public String getSparkJobDriverLogUrl(URI batchBaseUri, int batchId) throws IOEx
...
+
+            return driverHost;
+        });
     }
 
+    @Override
+    @NotNull
     public Observable<SimpleImmutableEntry<MessageInfoType, String>> getSubmissionLog() {
+        if (getConnectUri() == null) {
+            return Observable.error(new IOException("Spark job is not set up"));
```
Comment (djlan) [inline Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/spark/common/SparkBatchJob.java]: >Spark job is not set up [](start = 53, length = 23)

Same

File: Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/spark/common/SparkBatchJob.java [hunk 5]
```diff
@@ -719,6 +812,10 @@ public String getSparkDriverHost() throws IOException {
     }
 
     public boolean isActive() throws IOException {
+        if (getConnectUri() == null) {
+            throw new IOException("Spark job is not set up");
```
Comment (djlan) [inline Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/spark/common/SparkBatchJob.java]: >Spark job is not set up [](start = 35, length = 23)

Same

File: Utils/hdinsight-node-common/Test/java/com/microsoft/azure/hdinsight/spark/common/SparkBatchRemoteDebugJobScenario.java [hunk 6]
```diff
@@ -280,4 +273,15 @@ public void gettingSparkJobDriverLogURLObservableShouldBeEmpty() throws Throwabl
     public void mockSparkJobUriIsValidOrNot(String uriToCheck, String validOrNot) throws Throwable {
         doReturn(validOrNot.equals("valid")).when(debugJobMock).isUriValid(URI.create(uriToCheck));
     }
+
+    @And("^submit Spark job$")
```
Comment (djlan) [inline Utils/hdinsight-node-common/Test/java/com/microsoft/azure/hdinsight/spark/common/SparkBatchRemoteDebugJobScenario.java]: >@And("^submit Spark job$") [](start = 4, length = 26)

This test is for Spark debug job?

File: Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/spark/common/SparkBatchAzureSubmission.java (hunk 2 above)
Comment (wezhang) [reply Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/spark/common/SparkBatchAzureSubmission.java]: To fix it with changing the method  parameter to `CredentialsProvider`

File: Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/spark/common/SparkBatchRemoteDebugJob.java (hunk 3 above)
Comment (wezhang) [reply Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/spark/common/SparkBatchRemoteDebugJob.java]: fix it

File: Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/spark/common/SparkBatchJob.java [hunk 7]
```diff
@@ -219,14 +256,17 @@ public void setDelaySeconds(int delaySeconds) {
...
      *
      * @return the current instance for chain calling
      * @throws IOException the exceptions for networking connection issues related
      */
-    @Override
-    public ISparkBatchJob createBatchJob()
+    private SparkBatchJob createBatchJob()
             throws IOException {
+        if (getConnectUri() == null) {
+            throw new IOException("Spark job is not set up");
```
[the remark this answers is not in the document]
Comment (wezhang) [reply Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/spark/common/SparkBatchJob.java]: To fix it

File: Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/spark/common/SparkBatchJob.java [hunk 8]
```diff
@@ -252,20 +292,50 @@ public ISparkBatchJob createBatchJob()
...
-
-        if (deleteResponse.getCode() > 300) {
-            throw new UnknownServiceException(String.format(
-                    "Failed to stop spark remote debug job. error code: %d, reason: %s.",
-                    deleteResponse.getCode(), deleteResponse.getContent()));
-        }
+    public Observable<ISparkBatchJob> killBatchJob() {
+        return Observable.fromCallable(() -> {
+            if (getConnectUri() == null) {
+                throw new IOException("Spark job is not set up");
```
[the remark this answers is not in the document]
Comment (wezhang) [reply Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/spark/common/SparkBatchJob.java]: To fix it

File: Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/spark/common/SparkBatchJob.java [hunk 9]
```diff
@@ -292,6 +362,10 @@ protected String parseAmHostHttpAddressHost(@Nullable String driverHttpAddress)
      * @throws IOException exceptions in transaction
      */
     public String getState() throws IOException {
+        if (getConnectUri() == null) {
+            throw new IOException("Spark job is not set up");
```
[the remark this answers is not in the document]
Comment (wezhang) [reply Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/spark/common/SparkBatchJob.java]: To fix it

File: Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/spark/common/SparkBatchJob.java [hunk 10]
```diff
@@ -444,7 +519,11 @@ protected App getSparkJobYarnApplication(URI batchBaseUri, String applicationID)
      *
      * @return Yarn Application Attempt info Observable
      */
-    public Observable<AppAttempt> getSparkJobYarnCurrentAppAttempt() {
+    Observable<AppAttempt> getSparkJobYarnCurrentAppAttempt() {
+        if (getConnectUri() == null) {
+            return Observable.error(new IOException("Spark job is not set up"));
```
[the remark this answers is not in the document]
Comment (wezhang) [reply Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/spark/common/SparkBatchJob.java]: To fix it

File: Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/spark/common/SparkBatchJob.java [hunk 11]
```diff
@@ -483,7 +562,11 @@ protected App getSparkJobYarnApplication(URI batchBaseUri, String applicationID)
      *
      * @return The string pair Observable of Host and Container Id
      */
-    public Observable<SimpleImmutableEntry<URI, String>> getSparkJobYarnContainersObservable(@NotNull AppAttempt appAttempt) {
+    Observable<SimpleImmutableEntry<URI, String>> getSparkJobYarnContainersObservable(@NotNull AppAttempt appAttempt) {
+        if (getConnectUri() == null) {
+            return Observable.error(new IOException("Spark job is not set up"));
```
[the remark this answers is not in the document]
Comment (wezhang) [reply Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/spark/common/SparkBatchJob.java]: To fix it

File: Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/spark/common/SparkBatchJob.java [hunk 12]
```diff
@@ -636,38 +718,49 @@ public String getSparkJobDriverLogUrl(URI batchBaseUri, int batchId) throws IOEx
...
-     * @throws IOException exceptions for the driver host not found
+     * @return Spark driver node host observable
      */
     @Override
-    public String getSparkDriverHost() throws IOException {
-        String applicationId = this.getSparkJobApplicationId(this.getConnectUri(), this.getBatchId());
+    public Observable<String> getSparkDriverHost() {
+        return Observable.fromCallable(() -> {
+            if (getConnectUri() == null) {
+                throw new IOException("Spark job is not set up");
```
[the remark this answers is not in the document]
Comment (wezhang) [reply Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/spark/common/SparkBatchJob.java]: To fix it

File: Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/spark/common/SparkBatchJob.java (hunk 4 above)
Comment (wezhang) [reply Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/spark/common/SparkBatchJob.java]: To fix it

File: Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/spark/common/SparkBatchJob.java (hunk 5 above)
Comment (wezhang) [reply Utils/hdinsight-node-common/src/com/microsoft/azure/hdinsight/spark/common/SparkBatchJob.java]: To fix it

File: Utils/hdinsight-node-common/Test/java/com/microsoft/azure/hdinsight/spark/common/SparkBatchRemoteDebugJobScenario.java (hunk 6 above)
Comment (wezhang) [reply Utils/hdinsight-node-common/Test/java/com/microsoft/azure/hdinsight/spark/common/SparkBatchRemoteDebugJobScenario.java]: This test feature, Spark Job Spark Debug Job cases are mixed, so the naming here is for both.

**Pull request merged**